compilerlibs/Makefile.compilerlibs

```diff
@@ -261,14 +261,32 @@ MIDDLE_END_CMI=\
 OPTCOMP=$(MIDDLE_END) $(ASMCOMP)
 OPTCOMP_CMI=$(MIDDLE_END_CMI) $(ASMCOMP_CMI)
 
-TOPLEVEL=toplevel/genprintval.cmo toplevel/toploop.cmo \
-  toplevel/trace.cmo toplevel/topdirs.cmo toplevel/topmain.cmo
-TOPLEVEL_CMI=
+TOPLEVEL=toplevel/genprintval.cmo toplevel/byte/toploop.cmo \
+  toplevel/byte/trace.cmo toplevel/byte/topdirs.cmo toplevel/byte/topmain.cmo
+TOPLEVEL_CMI=toplevel/byte/toploop.cmi toplevel/byte/trace.cmi toplevel/byte/topdirs.cmi toplevel/byte/topmain.cmi
 
-OPTTOPLEVEL=toplevel/genprintval.cmo toplevel/opttoploop.cmo \
-  toplevel/opttopdirs.cmo toplevel/opttopmain.cmo
-OPTTOPLEVEL_CMI=
+OPTTOPLEVEL=toplevel/genprintval.cmo toplevel/native/toploop.cmo \
+  toplevel/native/topdirs.cmo toplevel/native/topmain.cmo
+OPTTOPLEVEL_CMI=toplevel/native/toploop.cmi toplevel/native/topdirs.cmi toplevel/native/topmain.cmi
 
+TOPLEVEL_SHARED_MLIS = toploop.mli topdirs.mli topmain.mli
+
+toplevel/%/toploop.cmi: toplevel/toploop.cmi
+	cp $< toplevel/toploop.mli $(@D)
```

Comment:
I can believe that this works, but it seems unnecessarily fishy to me. i.e. we don't declare one of the target.

Do you mind splitting this rule (as well as the next two) in two simpler copy rules?

Reply:
Indeed, it should work because the `cmi` is necessary more recent than the `mli` at this point, but...

The same is done [here](https://github.com/ocaml/ocaml/blob/7bbde5be870cdbd20281e3b7f0fcb2b36f755373/otherlibs/dynlink/Makefile#L199) ; but I'll test and change if there are no issues to two separate rules.

Reply:
Please tell me what you think of the new version; it's clearer and cleaner, but still slightly suspicious because it theoretically allows copying any mli/cmi from `toplevel/` to `toplevel/{byte,native}/` ; in practive it won't because of the rules generated by `make depends`, which we ensure are correct with the explicit `beforedepends` copy rule.

Reply:
We discussed this today with Louis. I was expecting that splitting the rule in two would be better, but it turns out that it leads to inconsistent assumptions. Given that the way the rules are currently written works and we do that in other parts of the Makefiles, we decided to leave them as is.